File: 2096-step-by-step-directions-from-a-binary-tree-node-to-another/2096-step-by-step-directions-from-a-binary-tree-node-to-another.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    string getDirections(TreeNode* root, int startValue, int destValue) {
        string srcDicrections="";
        int srcDepth=0;
        string directions="";
        int depth=0;
        directionsPlease(root,startValue,directions,depth,srcDicrections,srcDepth);

        string destDicrections="";
        int destDepth=0;
        directions="";
        depth=0;
        directionsPlease(root,destValue,directions,depth,destDicrections,destDepth);

        string ans="";
        int i=0;
        while(i<min(srcDicrections.length(),destDicrections.length())){
            if(srcDicrections[i]!=destDicrections[i]){
                break;
            }
            i++;
        }

        if(i<srcDicrections.length() && i<destDicrections.length()){
            int t=0;
            while(i<srcDicrections.length()){
                ans+='U';
                i++;
                t++;
            }

            int j=i-t;
            while(j<destDicrections.length()){
                ans+=destDicrections[j];
                j++;
            }
            return ans;
        }

        if(i<srcDicrections.length()){
            while(i<srcDicrections.length()){
                ans+='U';
                i++;
            }
            return ans;
        }

        if(i<destDicrections.length()){
            while(i<destDicrections.length()){
                ans+=destDicrections[i];
                i++;
            }
            return ans;
        }

        return ans;
    }

    void directionsPlease(TreeNode* root,int searchValue,string& curr,int& depth,string& directions,int& deep){
        if(!root){
            return;
        }

        if(root->val==searchValue){
            directions=curr;
            deep=depth;
            return;
        }

        curr+='L';
        depth+=1;
        directionsPlease(root->left,searchValue,curr,depth,directions,deep);
        curr.pop_back();
        curr+='R';
        directionsPlease(root->right,searchValue,curr,depth,directions,deep);
        curr.pop_back();
        depth-=1;
        return;
    }
};

File: 2096-step-by-step-directions-from-a-binary-tree-node-to-another/2096-step-by-step-directions-from-a-binary-tree-node-to-another.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public String getDirections(TreeNode root, int startValue, int destValue) {
        return combine2Paths(dfs(root,startValue,new StringBuilder()),dfs(root,destValue,new StringBuilder()));
    }

    public String dfs(TreeNode root, int val,StringBuilder path) {
        if (root == null) return "";
        if (root.val == val) return path.toString();
        String left = dfs(root.left,val,path.append("L"));
        if(left!="") return left;
        path.deleteCharAt(path.length()-1); 
        String right = dfs(root.right,val,path.append("R"));
        if(right!="") return right;
        path.deleteCharAt(path.length()-1); 
        return "";
    }

    
    public String combine2Paths(String s1,String s2) {
        int i=0;int j=0;
        StringBuilder s3 = new StringBuilder();
        while(i<s1.length() && j<s2.length()) {
            if(s1.charAt(i)==s2.charAt(j)) {
                i++;j++;
            } else break;
        }
        while(i<s1.length()) {s3.append("U");i++;}
        if(j<s2.length()) s3.append(s2.substring(j));
        return s3.toString();
    }
}